core/state_processor.go

```diff
@@ -98,23 +101,50 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 		allLogs = append(allLogs, receipt.Logs...)
 	}
 
+	// Overlay tree migration logic
+	migrdb := statedb.Database()
+	// TODO: avoid opening the preimages file here and make it part of, potentially, statedb.Database().
+	filePreimages, err := os.Open("preimages.bin")
+	if err != nil {
+		return nil, nil, 0, fmt.Errorf("opening preimage file: %s", err)
+	}
+	defer filePreimages.Close()
+	preimageSeek := migrdb.GetCurrentPreimageOffset()
+	if _, err := filePreimages.Seek(migrdb.GetCurrentPreimageOffset(), io.SeekStart); err != nil {
+		return nil, nil, 0, fmt.Errorf("seeking preimage file: %s", err)
+	}
+	fpreimages := bufio.NewReader(filePreimages)
+
 	// verkle transition: if the conversion process is in progress, move
 	// N values from the MPT into the verkle tree.
-	if statedb.Database().InTransition() {
+	if migrdb.InTransition() {
 		var (
 			now = time.Now()
 			tt  = statedb.GetTrie().(*trie.TransitionTrie)
 			mpt = tt.Base()
 			vkt = tt.Overlay()
 		)
 
-		accIt, err := statedb.Snaps().AccountIterator(mpt.Hash(), statedb.Database().GetCurrentAccountHash())
+		accIt, err := statedb.Snaps().AccountIterator(mpt.Hash(), migrdb.GetCurrentAccountHash())
 		if err != nil {
 			return nil, nil, 0, err
 		}
 		defer accIt.Release()
 		accIt.Next()
 
+		// If we're about to start with the migration process, we have to read the first account hash preimage.
+		if migrdb.GetCurrentAccountAddress() == (common.Address{}) {
+			var addr common.Address
+			if _, err := io.ReadFull(fpreimages, addr[:]); err != nil {
+				return nil, nil, 0, fmt.Errorf("reading preimage file: %s", err)
+			}
+			if crypto.Keccak256Hash(addr[:]) != accIt.Hash() {
+				return nil, nil, 0, fmt.Errorf("preimage file does not match account hash: %s != %s", crypto.Keccak256Hash(addr[:]), accIt.Hash())
+			}
+			preimageSeek += int64(len(addr))
+			migrdb.SetCurrentAccountAddress(addr)
+		}
```

Comment:
This code only runs once, on the first block of the sweeping phase.
It's basically setting up `SetCurrentAccountAddress(...)` to the first account. You can ignore the rest for now, so you can focus on that in the main loop.